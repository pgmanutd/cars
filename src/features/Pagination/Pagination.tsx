import React, { memo } from 'react';
import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import { Link as RouterLink } from 'react-router-dom';

import { useTranslate } from 'features/Translate';

const defaultPageNumber = 1;

const getPaginationLinksAttrs = ({
  currentPage,
  totalPage,
  href,
}: {
  currentPage: number;
  totalPage: number;
  href: string;
}) => {
  const first = {
    to: `${href}${defaultPageNumber}`,
    disabled: currentPage === defaultPageNumber,
  };

  const previous = {
    to: `${href}${currentPage - 1}`,
    disabled: currentPage === defaultPageNumber,
  };

  const next = {
    to: `${href}${currentPage + 1}`,
    disabled: currentPage === totalPage,
  };

  const last = {
    to: `${href}${totalPage}`,
    disabled: currentPage === totalPage,
  };

  return {
    first,
    previous,
    next,
    last,
  };
};

export interface PaginationProps {
  currentPage?: number;
  totalPage?: number;
  href?: string;
}

const Pagination: React.FC<PaginationProps> = ({
  currentPage = defaultPageNumber,
  totalPage = defaultPageNumber,
  href = `/?${new URLSearchParams({ page: '' })}`,
  ...restProps
}) => {
  const { translate } = useTranslate();

  const { first, previous, next, last } = getPaginationLinksAttrs({
    currentPage,
    totalPage,
    href,
  });

  return (
    <Box data-testid="Pagination" {...restProps} component="section">
      <Button component={RouterLink} color="primary" {...first}>
        {translate('features.Pagination.firstLinkText')}
      </Button>
      <Button component={RouterLink} color="primary" {...previous}>
        {translate('features.Pagination.previousLinkText')}
      </Button>
      <Typography variant="body2" display="inline" color="textPrimary">
        {translate('features.Pagination.description', {
          CURRENT_PAGE: currentPage.toString(),
          TOTAL_PAGE: totalPage.toString(),
        })}
      </Typography>
      <Button component={RouterLink} color="primary" {...next}>
        {translate('features.Pagination.nextLinkText')}
      </Button>
      <Button component={RouterLink} color="primary" {...last}>
        {translate('features.Pagination.lastLinkText')}
      </Button>
    </Box>
  );
};

export default memo(Pagination);
