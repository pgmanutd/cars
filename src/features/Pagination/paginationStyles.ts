import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

export const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      display: 'flex',
      justifyContent: 'center',
      alignItems: 'center',
    },
    text: {
      marginRight: theme.spacing(1),
      marginLeft: theme.spacing(1),
    },
  }),
);
