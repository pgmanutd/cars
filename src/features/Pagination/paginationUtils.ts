import { DEFAULT_PAGE_NUMBER } from './paginationConstants';

export const getPaginationLinksAttrs = ({
  currentPage,
  totalPage,
  href,
}: {
  currentPage: number;
  totalPage: number;
  href: string;
}) => {
  const first = {
    to: `${href}${DEFAULT_PAGE_NUMBER}`,
    disabled: currentPage === DEFAULT_PAGE_NUMBER,
  };

  const previous = {
    to: `${href}${currentPage - 1}`,
    disabled: currentPage === DEFAULT_PAGE_NUMBER,
  };

  const next = {
    to: `${href}${currentPage + 1}`,
    disabled: currentPage === totalPage,
  };

  const last = {
    to: `${href}${totalPage}`,
    disabled: currentPage === totalPage,
  };

  return {
    first,
    previous,
    next,
    last,
  };
};
