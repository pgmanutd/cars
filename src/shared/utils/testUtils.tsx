import React from 'react';
import { Router } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { render } from '@testing-library/react';

export const renderWithRouter = (
  ui: React.ReactElement,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  }: { route?: string; history?: MemoryHistory } = {},
) => ({
  ...render(<Router history={history}>{ui}</Router>),
  history,
});
