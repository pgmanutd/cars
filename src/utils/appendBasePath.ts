const appendBasePath = (basePath: string, path: string) =>
  `${basePath}?${path}`;

export default appendBasePath;
