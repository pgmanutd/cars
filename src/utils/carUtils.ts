import { Car } from 'features/CarInfo/types';

export const getCarName = ({ manufacturerName, modelName }: Car) =>
  `${manufacturerName} ${modelName}`;

export const getCarFeatures = (
  { stockNumber, mileage, fuelType, color }: Car,
  { stockNumberPrefix = '' }: { stockNumberPrefix?: string } = {
    stockNumberPrefix: '',
  },
) =>
  `${stockNumberPrefix}${stockNumber} - ${mileage?.number} ${mileage?.unit} - ${fuelType} - ${color}`;
