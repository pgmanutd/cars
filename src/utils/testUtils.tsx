import React from 'react';
import { Router, Route } from 'react-router-dom';
import { createMemoryHistory, MemoryHistory } from 'history';
import { HelmetProvider } from 'react-helmet-async';
import { render } from '@testing-library/react';

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    routerConfig: {
      path = '/',
      route = '/',
      history = createMemoryHistory({ initialEntries: [route] }),
    },
  }: {
    routerConfig: { path?: string; route?: string; history?: MemoryHistory };
  } = { routerConfig: {} },
) => ({
  ...render(
    <HelmetProvider>
      <Router history={history}>
        <Route
          path={path}
          render={routeProps => React.cloneElement(ui, { ...routeProps })}
        />
      </Router>
    </HelmetProvider>,
  ),
  history,
});
