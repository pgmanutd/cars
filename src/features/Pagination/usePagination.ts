import appendBasePath from 'utils/appendBasePath';

import useSetURLQuery from 'hooks/useSetURLQuery';

import { DEFAULT_PAGE_NUMBER } from './paginationConstants';

const usePagination = ({
  currentPage,
  totalPage,
  basePath,
  queryKey,
}: {
  currentPage: number;
  totalPage: number;
  basePath: string;
  queryKey: string;
}) => {
  const firstLink = useSetURLQuery({
    [queryKey]: DEFAULT_PAGE_NUMBER?.toString(),
  });
  const previousLink = useSetURLQuery({
    [queryKey]: (currentPage - 1)?.toString(),
  });
  const nextLink = useSetURLQuery({
    [queryKey]: (currentPage + 1)?.toString(),
  });
  const lastLink = useSetURLQuery({
    [queryKey]: totalPage?.toString(),
  });

  const first = {
    to: appendBasePath(basePath, firstLink),
    disabled: currentPage === DEFAULT_PAGE_NUMBER,
  };

  const previous = {
    to: appendBasePath(basePath, previousLink),
    disabled: currentPage === DEFAULT_PAGE_NUMBER,
  };

  const next = {
    to: appendBasePath(basePath, nextLink),
    disabled: currentPage === totalPage,
  };

  const last = {
    to: appendBasePath(basePath, lastLink),
    disabled: currentPage === totalPage,
  };

  return {
    first,
    previous,
    next,
    last,
  };
};

export default usePagination;
