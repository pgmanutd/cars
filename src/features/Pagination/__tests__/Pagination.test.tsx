import React from 'react';

import { renderWithProviders } from 'utils/testUtils';

import Pagination, { PaginationProps } from '../Pagination';

describe('<Pagination />', () => {
  const setup = ({ currentPage, totalPage }: PaginationProps) => {
    const { renderResult } = renderWithProviders(
      <Pagination currentPage={currentPage} totalPage={totalPage} />,
    );

    return {
      renderResult,
    };
  };

  test.each`
    currentPage | totalPage
    ${1}        | ${1}
    ${1}        | ${3}
    ${2}        | ${3}
    ${3}        | ${3}
  `(
    'should render the component and matches it against stored snapshot for currentPage: $currentPage, totalPage: $totalPage',
    ({ currentPage, totalPage }) => {
      const { renderResult } = setup({ currentPage, totalPage });

      expect(renderResult.asFragment()).toMatchSnapshot();
    },
  );
});
