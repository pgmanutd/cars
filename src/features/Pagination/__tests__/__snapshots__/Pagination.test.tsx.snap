// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination /> should render the component and matches it against stored snapshot for currentPage: 1, totalPage: 1 1`] = `
<DocumentFragment>
  <section
    class="MuiBox-root MuiBox-root-2"
    data-testid="Pagination"
  >
    <a
      aria-disabled="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-disabled Mui-disabled"
      href="/?page=1"
      role="button"
      tabindex="-1"
    >
      <span
        class="MuiButton-label"
      >
        First
      </span>
    </a>
    <a
      aria-disabled="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-disabled Mui-disabled"
      href="/?page=0"
      role="button"
      tabindex="-1"
    >
      <span
        class="MuiButton-label"
      >
        Previous
      </span>
    </a>
    <p
      class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextPrimary MuiTypography-displayInline"
    >
      Page 1 of 1
    </p>
    <a
      aria-disabled="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-disabled Mui-disabled"
      href="/?page=2"
      role="button"
      tabindex="-1"
    >
      <span
        class="MuiButton-label"
      >
        Next
      </span>
    </a>
    <a
      aria-disabled="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-disabled Mui-disabled"
      href="/?page=1"
      role="button"
      tabindex="-1"
    >
      <span
        class="MuiButton-label"
      >
        Last
      </span>
    </a>
  </section>
</DocumentFragment>
`;

exports[`<Pagination /> should render the component and matches it against stored snapshot for currentPage: 1, totalPage: 3 1`] = `
<DocumentFragment>
  <section
    class="MuiBox-root MuiBox-root-66"
    data-testid="Pagination"
  >
    <a
      aria-disabled="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-disabled Mui-disabled"
      href="/?page=1"
      role="button"
      tabindex="-1"
    >
      <span
        class="MuiButton-label"
      >
        First
      </span>
    </a>
    <a
      aria-disabled="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-disabled Mui-disabled"
      href="/?page=0"
      role="button"
      tabindex="-1"
    >
      <span
        class="MuiButton-label"
      >
        Previous
      </span>
    </a>
    <p
      class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextPrimary MuiTypography-displayInline"
    >
      Page 1 of 3
    </p>
    <a
      aria-disabled="false"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
      href="/?page=2"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiButton-label"
      >
        Next
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </a>
    <a
      aria-disabled="false"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
      href="/?page=3"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiButton-label"
      >
        Last
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </a>
  </section>
</DocumentFragment>
`;

exports[`<Pagination /> should render the component and matches it against stored snapshot for currentPage: 2, totalPage: 3 1`] = `
<DocumentFragment>
  <section
    class="MuiBox-root MuiBox-root-140"
    data-testid="Pagination"
  >
    <a
      aria-disabled="false"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
      href="/?page=1"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiButton-label"
      >
        First
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </a>
    <a
      aria-disabled="false"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
      href="/?page=1"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiButton-label"
      >
        Previous
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </a>
    <p
      class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextPrimary MuiTypography-displayInline"
    >
      Page 2 of 3
    </p>
    <a
      aria-disabled="false"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
      href="/?page=3"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiButton-label"
      >
        Next
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </a>
    <a
      aria-disabled="false"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
      href="/?page=3"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiButton-label"
      >
        Last
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </a>
  </section>
</DocumentFragment>
`;

exports[`<Pagination /> should render the component and matches it against stored snapshot for currentPage: 3, totalPage: 3 1`] = `
<DocumentFragment>
  <section
    class="MuiBox-root MuiBox-root-214"
    data-testid="Pagination"
  >
    <a
      aria-disabled="false"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
      href="/?page=1"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiButton-label"
      >
        First
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </a>
    <a
      aria-disabled="false"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
      href="/?page=2"
      role="button"
      tabindex="0"
    >
      <span
        class="MuiButton-label"
      >
        Previous
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </a>
    <p
      class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextPrimary MuiTypography-displayInline"
    >
      Page 3 of 3
    </p>
    <a
      aria-disabled="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-disabled Mui-disabled"
      href="/?page=4"
      role="button"
      tabindex="-1"
    >
      <span
        class="MuiButton-label"
      >
        Next
      </span>
    </a>
    <a
      aria-disabled="true"
      class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary Mui-disabled Mui-disabled"
      href="/?page=3"
      role="button"
      tabindex="-1"
    >
      <span
        class="MuiButton-label"
      >
        Last
      </span>
    </a>
  </section>
</DocumentFragment>
`;
